plugins/cros-ec/fu-cros-ec-usb-device.c

```diff
@@ -299,6 +314,248 @@ fu_cros_ec_usb_device_setup (FuDevice *device, GError **error)
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_cros_ec_usb_device_transfer_section (FuDevice *device,
+					FuFirmware *firmware,
+					FuCrosEcFirmwareSection *section,
+					gboolean smart_update,
+					GError **error)
+{
+	FuCrosEcUsbDevice *self = FU_CROS_EC_USB_DEVICE (device);
+	const guint8 * data_ptr = NULL;
+	guint32 section_addr = 0;
+	gsize data_len = 0;
+	gsize offset = 0;
+	g_autoptr(GBytes) img_bytes = NULL;
+
+	if (section == NULL) {
+		g_set_error_literal (error,
+				     G_IO_ERROR,
+				     G_IO_ERROR_INVALID_DATA,
+				     "no section information provided");
+		return FALSE;
+	}
+
+	section_addr = section->offset;
+	img_bytes = fu_firmware_get_image_by_idx_bytes (firmware,
+							section->image_idx,
+							error);
+	if (img_bytes == NULL) {
+		g_prefix_error (error, "failed to find section image: ");
+		return FALSE;
+	}
+
+	data_ptr = (const guint8 *)g_bytes_get_data (img_bytes, &data_len);
+	if (data_ptr == NULL || data_len != section->size) {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_INVALID_DATA,
+			     "image and section sizes do not match: image = %"
+			     G_GSIZE_FORMAT " bytes vs section size = %"
+			     G_GSIZE_FORMAT " bytes",
+			     data_len, section->size);
+		return FALSE;
+	}
+
+	if (smart_update) {
+		while (data_len != 0 && (data_ptr[data_len -1] == 0xff))
+			data_len--;
+		g_debug ("trimmed %" G_GSIZE_FORMAT " trailing bytes\n",
+			 section->size - data_len);
+	}
+
+	g_debug ("sending 0x%zx bytes to %#x\n", data_len, section_addr);
+	while (data_len > 0) {
+		gsize payload_size;
+		guint32 block_base;
+		fu_cros_ec_usb_block_info block_info;
+
+		/* prepare the header to prepend to the block */
+		block_info.image_bytes = img_bytes;
+		payload_size = MIN (data_len,
+				    self->targ.common.maximum_pdu_size);
+		block_base = GUINT32_TO_BE (section_addr);
+		block_info.ufh.block_size = GUINT32_TO_BE (payload_size +
+							   sizeof (struct update_frame_header));
+		block_info.ufh.cmd.block_base = block_base;
+		block_info.ufh.cmd.block_digest = 0;
+		block_info.offset = offset;
+		block_info.payload_size = payload_size;
+
+		if (!fu_device_retry (device,
+				      fu_cros_ec_usb_device_transfer_block,
+				      MAX_BLOCK_XFER_RETRIES, &block_info,
+				      error)) {
+			g_prefix_error (error,
+					"failed to transfer block, %"
+					G_GSIZE_FORMAT " to go: ", data_len);
+			return FALSE;
+		}
+		data_len -= payload_size;
+		offset += payload_size;
+		section_addr += payload_size;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_cros_ec_usb_device_send_done (FuDevice *device, GError **error)
+{
+	guint32 out = GUINT32_TO_BE (UPDATE_DONE);
+
+	/* send stop request, ignoring reply */
+	if (!fu_cros_ec_usb_device_do_xfer (FU_CROS_EC_USB_DEVICE (device),
+					    (guint8 *)&out, sizeof (out),
+					    (guint8 *)&out, 1,
+					    FALSE, NULL, error)) {
+		g_debug ("error on transfer of done. clearing error.\n");
```

Comment:
you shouldn't need to add a newline to the end here

Reply:
or the `.`

Reply:
Done.